### DISCUSSION TITLE
Middleware based on Route Parameter

### DISCUSSION THREAD

**Discussion by sanhardik** [fastapi/fastapi#4027] [Questions] [answered]
Hello, 
Is there a way to run a middleware based on a route variable ?

Consider an example:
I am trying to implement multiple endpoints. All of them has parameter 1 in common and I would like to be able to attach some custom information based on parameter 1's value and then let the endpoints process their own.

Endpoints: 
```
@app.get("/api/{p1}/abc")
def get_data(p1: str):
    if p1 == 'a':
        version = 'v1'
   if p1 == 'b':
        version = 'v2'
   if version == 'v1':
        return fake_data_1()
  elif version == 'v2':
        return fake_data_2()
  else:
        return None

@app.post("/api/{p1}/def")
def post_data(p1: str):
    if p1 == 'a':
        version = 'v1'
   if p1 == 'b':
        version = 'v2'
   if version == 'v1':
        store_fake_data_1()
  elif version == 'v2':
        store_fake_data_2()
```

Based on the value of p1, I would like to add custom information, which would be available to the  functions. 
As you can see in the example, if I could have processed the common functionality before, then I didnt need to have it in all the functions. 

Express - Node.js has similar implementation in http://expressjs.com/en/5x/api.html#app.param

Is it possible to do the same in FastAPI ?

Thanks

**Comment by ghandic** [comment]
You can create sub applications per version and mount them as separate versions of your API

**Reply by sanhardik** [reply]
Would you be able to point me to some doc's or examples for this ?

**Reply by ghandic** [reply]
https://fastapi.tiangolo.com/advanced/sub-applications/?h=sub

**Reply by ghandic** [reply]
If this helps solve your issue could you mark it as answered to help others? :)

**Reply by sanhardik** [reply]
Wouldn't this create code duplication ? 
The way it is done in that express library keeps the code in one place and allows to update the incoming req to add more data .

**Reply by ghandic** [reply]
You should keep your logic away from your routes as much as possible imo so your decoupled from using a rest interface.

Your functions want to be separate per version aswell to have minimal coupling

**Reply by sanhardik** [reply]
Sorry for the late response. 

I found this functionality in Flask https://flask.palletsprojects.com/en/2.0.x/patterns/urlprocessors/

Was wondering if FastAPI has something similar

**Comment by YuriiMotov** [comment, marked as answer]
I would move this common logic to dependency:

```python
async def get_version(p1: str):
    if p1 == 'a':
        return 'v1'
    elif p1 == 'b':
        return 'v2'
    else:
        return None


@app.get("/api/{p1}/abc")
def get_data(version: Annotated[str, Depends(get_version)]):
    if version == 'v1':
        return fake_data_1()
    elif version == 'v2':
        return fake_data_2()
    else:
        return None

@app.post("/api/{p1}/def")
def post_data(version: Annotated[str, Depends(get_version)]):
    if version == 'v1':
        store_fake_data_1()
    elif version == 'v2':
        store_fake_data_2()

```

Or
```python
async def get_version(p1: str):
    if p1 == 'a':
        return 'v1'
    elif p1 == 'b':
        return 'v2'
    else:
        return None


HANDLERS: dict[str, dict[str, Any]] = {
    "v1": {
        "fake_data": fake_data_1,
        "store_fake_data": store_fake_data_1
    },
    "v2": {
        "fake_data": fake_data_2,
        "store_fake_data": store_fake_data_1
    }
}

async def get_handlers(version: Annotated[str, Depends(get_version)]):
    handlers = HANDLERS.get(version, {})
    return handlers



@app.get("/api/{p1}/abc")
def get_data(handlers: Annotated[dict[str, Any], Depends(get_handlers)]):
    handler = handlers.get("fake_data")
    if not handler:
        return None
    return handler()

@app.post("/api/{p1}/def")
def post_data(handlers: Annotated[dict[str, Any], Depends(get_handlers)]):
    handler = handlers.get("store_fake_data")
    if not handler:
        pass
    handler()
```

You can further develop the idea by improving typing
